pkg/agent/batcher.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package agent
+
+import (
+	"context"
+
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	profilestorepb "github.com/parca-dev/parca/gen/proto/go/parca/profilestore/v1alpha1"
+	"google.golang.org/grpc"
+)
+
+type Batcher struct {
+	series      []*profilestorepb.RawProfileSeries
+	writeClient profilestorepb.ProfileStoreServiceClient
+	logger      log.Logger
+
+	mtx                *sync.RWMutex
+	lastProfileTakenAt time.Time
+	lastError          error
+}
+
+func NewBatcher(wc profilestorepb.ProfileStoreServiceClient) *Batcher {
+	return &Batcher{
+		series:      []*profilestorepb.RawProfileSeries{},
+		writeClient: wc,
+		mtx:         &sync.RWMutex{},
+	}
+}
+
+func (b *Batcher) loopReport(lastProfileTakenAt time.Time, lastError error) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+	b.lastProfileTakenAt = lastProfileTakenAt
+	b.lastError = lastError
+}
+
+func (b *Batcher) Run(ctx context.Context) error {
+	// TODO(Sylfrena): Make ticker duration configurable
+	const tickerDuration = 10 * time.Second
+
+	ticker := time.NewTicker(tickerDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+		}
+
+		err := b.batchLoop(ctx)
+		b.series = []*profilestorepb.RawProfileSeries{}
```

Comment:
This needs to be protected by the mutex because it can occur concurrently, I would say it would be best to do this at the end of the `batchLoop` function instead.